### ISSUE TITLE
Implement mikrotik resources with code generation

### ISSUE THREAD

**Issue by ddelnano** [ddelnano/terraform-provider-mikrotik#47]
Many of the mikrotik resource implementations are similar to each other and I think there is an opportunity to support a significant number of resources through code generation. This issue will track the work to test out replacing our manually written resources with autogenerated versions.

**Comment by maxim-nazarenko** [comment]
hi @ddelnano,

do you have thoughts regarding this?
Would it be sufficient to have a simple (or not) declaration in static yaml/json file or you want to fetch the fields from RouterOS instance?
I would like to help, but a quick research didn't lead me to any schema declaration for RouterOS objects except the official Wiki page for CLI.

**Comment by ddelnano** [comment]
Hey Maksym :wave:

I was thinking that the a static yaml/json file describing the RouterOS api would do. I'm not aware of any way to retrieve this information from RouterOS.

I started a branch [here](https://github.com/ddelnano/terraform-provider-mikrotik/tree/auto-generate-mikrotik-resources) that was going to be the beginning of this, but I haven't saw the idea fully through yet. There is [another repo](https://github.com/terra-farm/go-xen-api-client/blob/master/xenapi.json) that served as the inspiration for this idea, which you might find useful.

The goal would be to create both the terraform provider and acceptance test code. We will likely need a way to extend the auto generated code in the event it's difficult to have all resources fit the same mold or there need to be additional test cases supplied, etc.

Happy to help in any way that I can, but I don't have time to develop large pieces of this anytime soon.

**Comment by maxim-nazarenko** [comment]
thanks for you lightning response,

Without investing too much effort, I can imagine something like this:
* generate Terraform resources (possibly data objects) under [mikrotik](./mikrotik) folder
  * use key from declaration file to enable/disable resource `find`, `import`, etc. code generation
* generate necessary client methods under [client](./client) folder
  * use key from declaration file to provide Mikrotik API path to use (e.g. `/ip/dns/static`) and probably other configurations

Still not sure about Acceptance tests for Terraform and if it would be feasible to generate them.

Some questions off top of my head: what should happen when the autogenerated file is adjusted by hand (it was made more readable or efficient in sense of performance)?
Should this tool be used as one-time thing for scaffolding or it could be run multiple times for the same resource?

The schema for configuration file is a topic for discussion, but if there is no preference, we could start with something simple to not overengineer it.

Would appreciate if you share your thoughts about that.

**Comment by ddelnano** [comment]
>Still not sure about Acceptance tests for Terraform and if it would be feasible to generate them.

I think they should be relatively easy to generate. I don't think we need to generate all possible test cases. Just to provide enough test coverage for create, import and some update operations. As this project supports significantly more resources, I think it will be a huge asset to know that they work for basic operations.

> what should happen when the autogenerated file is adjusted by hand (it was made more readable or efficient in sense of performance)?

I think the autogenerated files should never be tweaked by hand. I think it's unlikely we will need to make something more efficient, but I do foresee situations with the terraform acceptance test generation where we may want to add additional test cases outside of the typical create, update and import tests to address bug fixes or special test cases.

I could see this working by allowing the code generation to concatenate (include) additional go template or raw files. I am picturing it would have a directory structure as seen below and the generation code could include a go template file at the end of its templated content.

```
mikrotik/mirotik_dns_resource_test.go
# generate code would append contents of mirkotik/additional_tests/mikrotik_dns_resource_test.tmpl 
# to mikrotik/mirotik_dns_resource_test.go
mirkotik/additional_tests/mikrotik_dns_resource_test.tmpl
```

> Should this tool be used as one-time thing for scaffolding or it could be run multiple times for the same resource? 

I think we should strive for ensuring that the code generation could run on every commit while keeping the "git status" of the project clean.

I don't have any strong preference for the template file, but I imagine I will be more opinionated once we see something that provides the necessary toggles for an initial implementation. But I agree that we should start basic and to avoid over-engineering it.

**Comment by maxim-nazarenko** [comment]
hey @ddelnano, 
sorry for a delayed reply (russian shit still arrives to our country).

At the moment I have rough implementation with 2 possible approaches.
The implementation does not produce tests and doesn't handle all possible Terraform properties, but they will be easy to add later.

1. Binary takes all necessary options via arguments
// client/dns.go
```go
//go:generate gen -struct=DnsRecord -requiredFields=name,address -omitFields=id -optionalFields=ttl -computedFields=ttl
type DnsRecord struct {
	Id      string `mikrotik:".id"`
	Name    string `mikrotik:"name"`
	Ttl     int    `mikrotik:"ttl,ttlToSeconds"`
	Address string `mikrotik:"address"`
}
```

2. Using struct tags (the struct name also could be discovered automatically, but I have to invest some time to find a way to do it)
```go
//go:generate gen -struct=DnsRecord
type DnsRecord struct {
	Id      string `mikrotik:".id"`
	Name    string `mikrotik:"name" gen:"name,required"`
	Ttl     int    `mikrotik:"ttl,ttlToSeconds" gen:"ttl,optional,computed"`
	Address string `mikrotik:"address" gen:"address,required"`
}
```

The second looks clear and explicit to me, but I would like to hear you opinion on that.

**Comment by maxim-nazarenko** [comment]
@ddelnano 
>hope you and your friends and family are able to stay safe :(

we are trying to do so, thank you

>Just to confirm the structs are inputs to the go:generate command and not outputs?

Yes, that's right. Here is a [gist](https://gist.github.com/maxim-nazarenko/f4fdca6f9caed5009b933fd6e0ecc326) with example of the output after running `go generate` for `DnsRecord` resource.

**Comment by maksym-nazarenko** [comment]
I'm resuming work on this issue, but it's 99% that I will open another PR.
The reasons are:
1.  This one is far behind master branch and I don't want to spend another hour merging it.
2. We are moving to new `PluginFramework` approach so I have to refactor it anyway.
3. I want to introduce this feature into the project as `utility` command to generate new resources with further manual review (probably, when it becomes stable, we could do `go generate` and commit immediately)

I will post new PR link here, once it is ready at least as `draft`
